### PR TITLE
Extract moveaxis and normalize_axis_tuple from einsum.

### PR DESCRIPTION
**Pull request by fukatani**
Extracted `moveaxis` as numpy>=1.11 API.
Also I want to use `normalize_axis_tuple` in `moveaxis` and `einsum`, I transport it to `cupy.core`.

### CODE REVIEW

File: cupy/manipulation/transpose.py [hunk 1]
```diff
@@ -36,6 +36,30 @@ def swapaxes(a, axis1, axis2):
...
+            also be unique.
+
+    Returns:
+        cupy.ndarray:
+        Array with moved axes. This array is a view of the input array.
+
+    .. seealso:: :func:`numpy.moveaxis`
+
+    """
+    # TODO(okuta): check type
```
Comment (unnonouno) [inline cupy/manipulation/transpose.py]: write your name

File: cupy/manipulation/transpose.py [hunk 2]
```diff
@@ -36,6 +36,30 @@ def swapaxes(a, axis1, axis2):
...
 
 
+def moveaxis(a, source, destination):
+    """Moves axes of an array to new positions.
+
+    Other axes remain in their original order.
+
+    Args:
+        a (cupy.ndarray): Array whose axes should be reordered.
+        source (int or sequence of int):
```
Comment (unnonouno) [inline cupy/manipulation/transpose.py]: Is int supported?

File: cupy/manipulation/transpose.py [hunk 3]
```diff
@@ -36,6 +36,30 @@ def swapaxes(a, axis1, axis2):
...
+            Original positions of the axes to move. These must be unique.
+        destination (int or sequence of int):
```
Comment (unnonouno) [inline cupy/manipulation/transpose.py]: ditto

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -2166,6 +2166,48 @@ cpdef ndarray asfortranarray(ndarray a, dtype=None):
...
+            raise numpy.AxisError('axis {} is out of bounds for array of '
+                                  'dimension {}'.format(ax, ndim))
+    return tuple((ax % ndim) for ax in axis)
+
+
+cpdef ndarray moveaxis(ndarray a, vector.vector[Py_ssize_t] source,
+                       vector.vector[Py_ssize_t] destination):
+
+    if len(source) != len(destination):
+        raise ValueError('`source` and `destination` arguments must have '
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Add test for this case

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -2166,6 +2166,48 @@ cpdef ndarray asfortranarray(ndarray a, dtype=None):
...
+    return False
+
+
+def normalize_axis_tuple(vector.vector[Py_ssize_t] axis, ndim):
+    """Normalizes an axis argument into a tuple of non-negative integer axes.
+
+    """
+    for ax in axis:
+        if ax >= ndim or ax < -ndim:
+            raise numpy.AxisError('axis {} is out of bounds for array of '
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Add test for this case

**Comment by unnonouno** [comment]
jenkis test this please

File: cupy/core/core.pyx (hunk 5 above)
Comment (unnonouno) [reply cupy/core/core.pyx]: Please use `cupy.core.core._AxisError` because old numpy raises `ValueError`.

**Comment by unnonouno** [comment]
jenkins test this please

File: cupy/manipulation/transpose.py (hunk 2 above)
Comment (fukatani) [reply cupy/manipulation/transpose.py]: Sorry, I fixed `normalize_axis_tuple` to support `int` and added the test.

**Comment by fukatani** [comment]
Thank you!
I addressed your comments.

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
Please check travis result (flake8)

**Comment by fukatani** [comment]
Sorry, I fixed coding style.

**Comment by unnonouno** [comment]
jenkins, test this please

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
Oh, the test cannot pass with the old numpy. The old NumPy raises different type of exceptions 
 when an invalid axis is found. We made `IndexOrValueError` to treat this problem. Please check with the old numpy.

**Comment by fukatani** [comment]
I checked with old numpy in current implemetation.
Indeed `cupy.moveaxis` raises `IndexOrValueError`,
But test was failed.


```======================================================================
FAIL: test_moveaxis_invalid1 (cupy_tests.manipulation_tests.test_transpose.TestTranspose)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ryo/workspace/github/cupy/cupy/testing/helper.py", line 478, in test_func
    accept_error=accept_error)
  File "/home/ryo/workspace/github/cupy/cupy/testing/helper.py", line 59, in _check_cupy_numpy_error
    self.fail(msg)
AssertionError: Different types of errors occurred

cupy
Traceback (most recent call last):
  File "/home/ryo/workspace/github/cupy/cupy/testing/helper.py", line 457, in test_func
    impl(self, *args, **kw)
  File "/home/ryo/workspace/github/cupy/tests/cupy_tests/manipulation_tests/test_transpose.py", line 29, in test_moveaxis_invalid1
    return xp.moveaxis(a, [0, 1], [1, 3])
  File "/home/ryo/workspace/github/cupy/cupy/manipulation/transpose.py", line 60, in moveaxis
    return core.moveaxis(a, source, destination)
  File "cupy/core/core.pyx", line 2193, in cupy.core.core.moveaxis
  File "cupy/core/core.pyx", line 2195, in cupy.core.core.moveaxis
  File "cupy/core/core.pyx", line 2187, in cupy.core.core.normalize_axis_tuple
cupy.core.core.IndexOrValueError: axis 3 is out of bounds for array of dimension 3

numpy
Traceback (most recent call last):
  File "/home/ryo/workspace/github/cupy/cupy/testing/helper.py", line 469, in test_func
    impl(self, *args, **kw)
  File "/home/ryo/workspace/github/cupy/tests/cupy_tests/manipulation_tests/test_transpose.py", line 29, in test_moveaxis_invalid1
    return xp.moveaxis(a, [0, 1], [1, 3])
  File "/home/ryo/workspace/cupy-env/lib/python3.5/site-packages/numpy/core/numeric.py", line 1535, in moveaxis
    destination = _validate_axis(destination, a.ndim, 'destination')
  File "/home/ryo/workspace/cupy-env/lib/python3.5/site-packages/numpy/core/numeric.py", line 1472, in _validate_axis
    argname)
ValueError: invalid axis for this array in `destination` argument
```

This result means that `numpy_cupy_raises` regards `ValueError` and `IndexAndValueError` are different.
Because `type(IndexOrValueError()) is not type(ValueError())`.

**Comment by fukatani** [comment]
`cupy.roll` also raises `_AxisError`, its test is written as follows.
```python
    @testing.for_all_dtypes()
    @testing.with_requires('numpy>=1.13')
    @testing.numpy_cupy_raises()
    def test_roll_invalid_negative_axis1(self, xp, dtype):
        x = testing.shaped_arange((5, 2), xp, dtype)
        return xp.roll(x, 1, axis=-3)

    @testing.for_all_dtypes()
    def test_roll_invalid_negative_axis2(self, dtype):
        x = testing.shaped_arange((5, 2), cupy, dtype)
        with self.assertRaises(cupy.core.core._AxisError):
            return cupy.roll(x, 1, axis=-3)
```

But I want helper which handles `IndexOrValueError` well.

**Comment by unnonouno** [comment]
Yes, it's ok.

**Comment by fukatani** [comment]
Thanks, I fixed the tests.

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by unnonouno** [comment]
Thank you very much! Nice work.

**Pull request merged**